Add page to view all builds/resource versions downstream/upstream from a root resource version

chenbh opened the pull request:
## Changes proposed by this PR

closes #6904 

## Notes to reviewer

- it was too hard to adapt the pipeline graph rendering to this view, so we ended up using a separate renderer
- view is scoped to a single pipeline
- build -> resource version -> build relationship is established via ~`passed` constraints (might experiment with the more computationally expensive~ `versionmd5` based query)

## Release Note

- Disabled by default since computing causality for large datasets can be expensive, use `--enable-resource-causality` or `$CONCOURSE_ENABLE_RESOURCE_CAUSALITY=true` to enable the web UI and API endpoint.
    - Most datasets ([like the merge commit for this PR](https://ci.concourse-ci.org/teams/main/pipelines/concourse/resources/concourse/causality/1014864917/downstream)) have < 100 builds and/or resource versions and take < 100ms, but it's possible for some "slow paced" resource versions (i.e. very infrequent new versions) to generate extremely large datasets
    - There is an automatic cutoff at 5000 builds or 25000 resource versions. On our deployment, the call for our [slowest paced resource](https://ci.concourse-ci.org/teams/main/pipelines/concourse/resources/oci-build-task?filter=digest%3Asha256%3A77eee31c5b5f8c426bd4b391fb00c7eaa734593b54b51edfeb22960192056e0f&filter=tag%3A0.7.0) took about ~7 seconds to process, most of which is spent in the DB query
- The causality page can be navigated to from the resource page
![Screen Shot 2021-06-03 at 11 37 08 AM](https://user-images.githubusercontent.com/17577073/120672162-0ed9ee80-c460-11eb-9a63-0a6f6a636bbf.png)
- The causality page displays all the builds and resource versions that was generated from (downstream) or resulted in (upstream) the creation of a particular resource version
![Screen Shot 2021-06-03 at 11 25 03 AM](https://user-images.githubusercontent.com/17577073/120671085-0a610600-c45f-11eb-82dd-1252f0d87da2.png)
- The downstream graph will put the root resource version on the left whereas the upstream graph will put it on the right
- It takes into account all the intermediate resource versions when computing the final graph. In the picture above, while the resource page only shows that `git version: 123` is a direct input to integrate `#4 & #5`, there is also an indirect link from `git version: 123` -> `test #19` -> ... -> `intermediate-3 version:123` -> `integrate #6 & #6.1`

Commit:
rework causality endpoint

[#6904]

1. use the build_pipes table to figure out list of builds that were
   generated as result of a particular resource version
1. figure out other resource versions that were used as input to those
   builds
1. figure out what resource versions were generated as a result of those
   builds
1. combine those data into a tree struct

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
render causality graph in ui

[#6904]

- redid api endpoint to query by resource_config_version_id instead of
  arbitrary filters
- expanding a resoruce version on the resource page will show the
  causality graph instead of the build inputs/outputs
- use something called a force graph (same one used by caded) to render
  it, i suspect it would look a lot nicer if we used a DAG

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

Commit:
web: text representation of causality

[#6904]

simple experiment to see how it would look like to render the graph
using text only.

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

Commit:
Revert "web: text representation of causality"

[#6904]

we want to iterate more on the graph representation, but wanted to keep
the text representation in git history

Co-authored-by: Alex Suraci <suraci.alex@gmail.com>
Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
cleanup api response

[#6904]

resource_config_version_id is useful for constructing nodes (if we
ignore builds)

having the proper version instead of a json string is nicer to render

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

Commit:
make the graph directional

[#6904]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Alex Suraci <suraci.alex@gmail.com>

Commit:
add upstream causality

before, it would only figure out the tree of resources and builds that
were generated as a result of the root resource version, now it also figures
out the builds and resources that resulted in the creation of the root
resource version.

the api response now returns 2 separate trees joined at the root node,
the root.outputOf represents the upstream and root.inputTo is the
downstream.

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
add upstream builds and resources to graph

currently dumping it in graphviz DOT format until we figure out how to
display it nicely

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
atc: split causality into upstream and downstream

as well as making the json response interchangable

[concourse/concourse#6904]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
web: add dedicated causality view

navigatable from the expanded resource versions tab. it's pretty baren
right now, but next goal is to actually implement the rendering

[concourse/concourse#6904]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
use dagre-d3 to render causality

- the causality tree now uses jobs and resources as the nodes (as
  opposed to builds and resource versions)
- multiple builds/resource versions for the same job/resource are
  collapsed into a single node

[concourse/concourse#6904]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
polish up the causality page

- add breadcrumbs to pipeline/resource/resource+version in causality
  page
- add tooltips and make it look like a proper button in the resources
  page

[concourse/concourse#6904]

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
fix elm tests

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
regenerate fakes

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
web: backfill tests

TODO:
- test that parsing tree -> graph works correctly
- test that parsing graph -> DOT works correctly

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
switch to hpcc-js/wasm for rendering causality

[concourse/concours#6904]

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
simplify graph representation

record -> variant

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
add styling to causality graph

- vender in Graph.DOT to support setting a node's label as an HTML label
- render each node as a table to get styling on each individual line
- add href to build and resource page to job and resource nodes
  respectively
- include build status in the api response so we can color the boxes
  accordingly

[concourse/concourse#6904]

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
add tests

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented on web/elm/src/Causality/Causality.elm:
```diff
@@ -0,0 +1,584 @@
+
+attributes : List ( String, String ) -> String
+attributes =
+    List.map (\( k, v ) -> k ++ "=\"" ++ v ++ "\"") >> String.join " "
```
They value (and probably key to be safe, though we're the ones who pick the keys) need to be escaped since they are encoded in the pseudo-html, and can break things otherwise, e.g.

```elm
escape : String -> String
escape =
    String.replace "<" "&lt;"
        >> String.replace ">" "&gt;"
        >> String.replace "&" "&amp;"
        >> String.replace "\"" "&quot;"
        >> String.replace "'" "&#039;"


attributes : List ( String, String ) -> String
attributes =
    List.map (\( k, v ) -> escape k ++ "=\"" ++ escape v ++ "\"") >> String.join " "
```

Note that this no longer requires `Json.Encode.string` to avoid the value containing a quote since we escape it properly

---
Actually, maybe we want an `htmlAttributes` that does this escaping, since this escaping isn't valid for "normal" attributes - but I suppose it doesn't really matter too much

chenbh replied:
> Actually, maybe we want an htmlAttributes that does this escaping

Not quite sure I know what this means. Either way, I think we'll need to keep this function because even the cell text also needs to be escaped (i.e. `<TD>&lt;special&amp;name&gt;</TD>`)

aoldershaw replied:
Right, I guess I just meant the way that characters get escaped when they are just regular attributes (as opposed to "HTML" attributes) is different. For instance, `"` should become `&quot;` in HTML, but `\"` in regular attributes.

That said, we don't currently really *need* to escape anything in the regular attributes, though, so that's why I say it doesn't matter too much

aoldershaw replied:
But in any event, c85181d215df0676700d7ce89af89026d722a5c6 isn't sufficient for making things not break - e.g. consider https://ci.concourse-ci.org/teams/main/pipelines/aidan/resources/mock/causality/942884455/downstream

<img width="774" alt="Screen Shot 2021-06-04 at 1 34 44 PM" src="https://user-images.githubusercontent.com/26583442/120841552-a2cbb900-c539-11eb-8ded-12d5512ff514.png">

Commit:
fix multi-key resource versions

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

cirocosta commented on atc/db/resource.go:
```diff
@@ -953,3 +956,244 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+	if err != nil {
+		return root, false, err
+	}
+	err = r.getCausalityResourceVersions(buildIDs, &root, updateInput, updateOutput)
```
whoopsy? looks like it's unchecked 😁  not sure if intentional/not

Commit:
properly escape all user input in causality page

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented on web/assets/css/svg.less:
```diff
@@ -1,7 +1,39 @@
 
 svg {
-  .node rect { fill: @grey100; }
+  position: absolute;
```
This breaks resource icons on the resource page
<img width="434" alt="Screen Shot 2021-06-04 at 12 01 00 PM" src="https://user-images.githubusercontent.com/26583442/120830535-9725c580-c52c-11eb-9535-d931f64cb4ad.png">

chenbh replied:
Should be fixed now

Commit:
web: reset pipelineLocator using the api response

model.pipelineLocator is set via the flags, which is in turn set by the
route. this can cause problems with pipelines that have special
characters in them since the parser will url encode them, which causes
the pipelineLocator.pipelineName to be `%3c', but the api response will
be for pipeline name '<'. this causes the pipeline lookup to fail which
ultimately results in a missing pipeline breadcrumb

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
fix css styling

- let's only set `position:absolute` on the graph svgs instead of all
  svgs
- if the breadcrumbs are too long, it should overflow instead of pushing
  the login/out button off the screen

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
cleanup atc

- get rid of the old Pipeline.Causality
- slightly better comments

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
fix yarn analyse

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
regenerate fakes

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented on atc/db/resource.go:
```diff
@@ -953,3 +956,243 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+		From("build_resource_config_version_inputs i").
+		Join("resources r ON r.id = i.resource_id").
+		Join("resource_config_versions rcv ON rcv.version_md5 = i.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id").
+		Where(sq.Eq{"i.build_id": buildIDs}).
```
Is it feasible to do this without a `WHERE build_id IN (....)` condition? I imagine that the number of build ids could be pretty huge, so doing a `JOIN` would probably be more scalable.

e.g.

```sql
WITH RECURSIVE build_ids AS
(
...
)
SELECT ...
FROM build_resource_config_version_inputs i
JOIN ...
JOIN build_ids bi ON i.build_id = bi.id
...
UNION ALL
SELECT ...
FROM build_resource_config_version_outputs o
JOIN ...
JOIN build_ids bi ON o.build_id = bi.id
...
```

chenbh replied:
You mean by embedding the `WITH RECURSIVE` query inside each of the queries for builds, build inputs, and build outputs? Is postgres capable of "learning" from the result of previous executions and make more informed plans?

My main concern is that the `WITH RECURSIVE` query should be the most expensive one, so we should only run it once and use the results of that in the 3 cheaper queries.

chenbh replied:
Or is it possible to do this in a transaction, store the result of the `WITH RECURSIVE` as a temp view/table, and then refer to that table in subsequent queries in the same transaction?

aoldershaw replied:
Well at least the inputs and outputs can be a single query with the `UNION ALL`, but yeah I was forgetting about the "build info" query

I know that the list passed to a `WHERE ... IN (...)`  has a limit of 65000 or so elements - but at that point, the performance is already probably pretty bad. Temp table is an option but feels pretty heavy weight, especially for the common case which shouldn't be too crazy.

Tagging @clarafu - WDYT?

chenbh replied:
I just want to point out that it needs to execute different functions depending on if it's processing the inputs or outputs. Which should be easy to do by adding a `'input' AS type` to the select statement and then special casing on that in the code

https://github.com/concourse/concourse/blob/a7b862ea40fa15d308ff383784b46d9f5fa58345/atc/db/resource.go#L1176-L1183

aoldershaw replied:
re: https://github.com/concourse/concourse/pull/7125#issuecomment-855015082:

It looks like the recursive query is actually quite fast

```
CTE Scan on build_ids  (cost=694.57..709.99 rows=771 width=8) (actual time=1.119..196.107 rows=17056 loops=1)
  CTE build_ids
    ->  Recursive Union  (cost=8.59..694.57 rows=771 width=8) (actual time=1.118..191.759 rows=17056 loops=1)
          ->  Unique  (cost=8.59..8.59 rows=1 width=8) (actual time=1.117..1.278 rows=661 loops=1)
                ->  Sort  (cost=8.59..8.59 rows=1 width=8) (actual time=1.116..1.173 rows=661 loops=1)
                      Sort Key: i.build_id
                      Sort Method: quicksort  Memory: 55kB
                      ->  Index Scan using build_inputs_resource_versions_idx on build_resource_config_version_inputs i  (cost=0.56..8.58 rows=1 width=8) (actual time=0.022..0.973 rows=661 loops=1)
                            Index Cond: ((resource_id = 101) AND (version_md5 = '40fa80af72b9c40a7b218fd3fb5393c0'::text))
          ->  HashAggregate  (cost=66.29..67.06 rows=77 width=8) (actual time=26.320..26.764 rows=2342 loops=7)
                Group Key: bp.to_build_id
                ->  Nested Loop  (cost=0.98..66.09 rows=77 width=8) (actual time=0.566..22.632 rows=21790 loops=7)
                      Join Filter: (bi.build_id = i_1.build_id)
                      ->  Nested Loop  (cost=0.42..48.73 rows=14 width=24) (actual time=0.557..6.702 rows=2535 loops=7)
                            ->  WorkTable Scan on build_ids bi  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.463 rows=2437 loops=7)
                            ->  Index Only Scan using build_pipes_from_build_id_to_build_id_uniq on build_pipes bp  (cost=0.42..4.84 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=17056)
                                  Index Cond: (from_build_id = bi.build_id)
                                  Heap Fetches: 663
                      ->  Index Only Scan using build_resource_config_version_inputs_uniq on build_resource_config_version_inputs i_1  (cost=0.56..1.17 rows=6 width=8) (actual time=0.003..0.004 rows=9 loops=17745)
                            Index Cond: (build_id = bp.from_build_id)
                            Filter: (resource_id <> 101)
                            Rows Removed by Filter: 0
                            Heap Fetches: 12926
Planning time: 0.469 ms
Execution time: 197.186 ms
```

On the other hand, the massing `WHERE build_id IN (...)` condition seems to be really slow - the `EXPLAIN ANALYZE` has been running for over 3 minutes at this point, still not complete.

So, doing it as a `JOIN` seems like a better option, even if it means needing to run the query twice. For this query:

```sql
WITH RECURSIVE build_ids AS (
		SELECT DISTINCT i.build_id
			FROM build_resource_config_version_inputs i
			WHERE i.resource_id=101 AND i.version_md5='40fa80af72b9c40a7b218fd3fb5393c0'
	UNION ALL
		SELECT DISTINCT bp.to_build_id AS build_id
		FROM build_ids bi
		INNER JOIN build_pipes bp ON bi.build_id = bp.from_build_id
		INNER JOIN build_resource_config_version_inputs i ON i.build_id = bi.build_id
		WHERE i.resource_id!=101
)
SELECT r.id, rcv.id, r.name, rcv.version, i.build_id, 'input' as type
FROM build_resource_config_version_inputs i
JOIN resources r ON r.id = i.resource_id
JOIN resource_config_versions rcv ON rcv.version_md5 = i.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id
JOIN build_ids bi ON i.build_id = bi.build_id
UNION ALL
SELECT r.id, rcv.id, r.name, rcv.version, o.build_id, 'output' as type
FROM build_resource_config_version_outputs o
JOIN resources r ON r.id = o.resource_id
JOIN resource_config_versions rcv ON rcv.version_md5 = o.version_md5 AND rcv.resource_config_scope_id = r.resource_config_scope_id
JOIN build_ids bi ON o.build_id = bi.build_id
```

...here's the `EXPLAIN ANALYZE`:

```
Append  (cost=1535.45..54855.85 rows=3 width=113) (actual time=3.821..1212.134 rows=134774 loops=1)
  CTE build_ids
    ->  Recursive Union  (cost=8.59..694.57 rows=771 width=8) (actual time=1.654..215.981 rows=17090 loops=1)
          ->  Unique  (cost=8.59..8.59 rows=1 width=8) (actual time=1.653..1.983 rows=661 loops=1)
                ->  Sort  (cost=8.59..8.59 rows=1 width=8) (actual time=1.652..1.826 rows=661 loops=1)
                      Sort Key: i_1.build_id
                      Sort Method: quicksort  Memory: 55kB
                      ->  Index Scan using build_inputs_resource_versions_idx on build_resource_config_version_inputs i_1  (cost=0.56..8.58 rows=1 width=8) (actual time=0.026..1.472 rows=661 loops=1)
                            Index Cond: ((resource_id = 101) AND (version_md5 = '40fa80af72b9c40a7b218fd3fb5393c0'::text))
          ->  HashAggregate  (cost=66.29..67.06 rows=77 width=8) (actual time=28.771..29.864 rows=2347 loops=7)
                Group Key: bp.to_build_id
                ->  Nested Loop  (cost=0.98..66.09 rows=77 width=8) (actual time=0.472..24.682 rows=21834 loops=7)
                      Join Filter: (bi_2.build_id = i_2.build_id)
                      ->  Nested Loop  (cost=0.42..48.73 rows=14 width=24) (actual time=0.462..7.386 rows=2540 loops=7)
                            ->  WorkTable Scan on build_ids bi_2  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.441 rows=2441 loops=7)
                            ->  Index Only Scan using build_pipes_from_build_id_to_build_id_uniq on build_pipes bp  (cost=0.42..4.84 rows=1 width=16) (actual time=0.002..0.002 rows=1 loops=17090)
                                  Index Cond: (from_build_id = bi_2.build_id)
                                  Heap Fetches: 697
                      ->  Index Only Scan using build_resource_config_version_inputs_uniq on build_resource_config_version_inputs i_2  (cost=0.56..1.17 rows=6 width=8) (actual time=0.003..0.005 rows=9 loops=17779)
                            Index Cond: (build_id = bp.from_build_id)
                            Filter: (resource_id <> 101)
                            Rows Removed by Filter: 0
                            Heap Fetches: 13250
  ->  Nested Loop  (cost=840.88..39800.98 rows=2 width=113) (actual time=3.821..1111.347 rows=127307 loops=1)
        ->  Hash Join  (cost=840.33..4550.73 rows=4254 width=63) (actual time=3.704..432.316 rows=127835 loops=1)
              Hash Cond: (i.resource_id = r.id)
              ->  Nested Loop  (cost=0.56..3652.47 rows=4254 width=45) (actual time=1.679..379.326 rows=127835 loops=1)
                    ->  CTE Scan on build_ids bi  (cost=0.00..15.42 rows=771 width=8) (actual time=1.656..221.465 rows=17090 loops=1)
                    ->  Index Only Scan using build_resource_config_version_inputs_uniq on build_resource_config_version_inputs i  (cost=0.56..4.66 rows=6 width=45) (actual time=0.006..0.008 rows=7 loops=17090)
                          Index Cond: (build_id = bi.build_id)
                          Heap Fetches: 11169
              ->  Hash  (cost=807.12..807.12 rows=2612 width=22) (actual time=1.983..1.983 rows=2612 loops=1)
                    Buckets: 4096  Batches: 1  Memory Usage: 174kB
                    ->  Seq Scan on resources r  (cost=0.00..807.12 rows=2612 width=22) (actual time=0.011..1.488 rows=2612 loops=1)
        ->  Index Scan using resource_config_scope_id_and_version_md5_unique on resource_config_versions rcv  (cost=0.56..8.28 rows=1 width=92) (actual time=0.005..0.005 rows=1 loops=127835)
              Index Cond: ((resource_config_scope_id = r.resource_config_scope_id) AND (version_md5 = i.version_md5))
  ->  Nested Loop  (cost=1.25..14360.26 rows=1 width=113) (actual time=0.048..85.249 rows=7467 loops=1)
        ->  Nested Loop  (cost=0.70..3024.50 rows=1368 width=63) (actual time=0.032..50.585 rows=8122 loops=1)
              ->  Nested Loop  (cost=0.42..2545.01 rows=1368 width=45) (actual time=0.019..36.930 rows=8122 loops=1)
                    ->  CTE Scan on build_ids bi_1  (cost=0.00..15.42 rows=771 width=8) (actual time=0.001..3.806 rows=17090 loops=1)
                    ->  Index Only Scan using build_resource_config_version_outputs_uniq on build_resource_config_version_outputs o  (cost=0.42..3.26 rows=2 width=45) (actual time=0.002..0.002 rows=0 loops=17090)
                          Index Cond: (build_id = bi_1.build_id)
                          Heap Fetches: 185
              ->  Index Scan using resources_pkey on resources r_1  (cost=0.28..0.34 rows=1 width=22) (actual time=0.001..0.001 rows=1 loops=8122)
                    Index Cond: (id = o.resource_id)
        ->  Index Scan using resource_config_scope_id_and_version_md5_unique on resource_config_versions rcv_1  (cost=0.56..8.28 rows=1 width=92) (actual time=0.004..0.004 rows=1 loops=8122)
              Index Cond: ((resource_config_scope_id = r_1.resource_config_scope_id) AND (version_md5 = o.version_md5))
Planning time: 3.089 ms
Execution time: 1220.851 ms
```

clarafu replied:
Hmm ya I definitely get the concern but right now I can't think of an easy way around it. 🤔  I was also thinking iof we can just combine the queries so we don't have to pass around the list of builds but I don't think we can attach fetching the resource/version/build info in the recursive query. The parent query that runs in order to do the `UNION ALL` will be super slow if it tries to union on all the resource/version/build info rather than just doing the union on the builds and then we fetch out the info after. 

As an example, I tried running this against `hush-house` with approx 200 builds and it took around 800 ms to run. https://explain.depesz.com/s/8rfJ

I think if we can think of a better way to do it that is easy enough then yes we should probably go ahead with that , but for now I think this current approach is good enough for this first iteration!

aoldershaw replied:
@clarafu thoughts on just running the recursive query 2 (or 3) times, per https://github.com/concourse/concourse/pull/7125#discussion_r646657959? That last `EXPLAIN ANALYZE` is run on a resource version with 17000 builds and it doesn't get *too* slow

chenbh commented:
Fun fact, it took 8.6s to load `oci-build-task`, which is probably the most used resource version on our CI outside of time related ones like `oxygen-mask`. 

https://ci.concourse-ci.org/teams/main/pipelines/concourse/resources/oci-build-task/causality/591570659/downstream

clarafu commented on atc/db/resource.go:
```diff
@@ -953,3 +956,243 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+		SELECT DISTINCT bp.to_build_id AS build_id
+		FROM build_ids bi
+		INNER JOIN build_pipes bp ON bi.build_id = bp.from_build_id
+		INNER JOIN build_resource_config_version_inputs i ON i.build_id = bi.build_id
```
I ran this query against `hush-house` on a resource version that had approx 200 builds, and it seemed to take around 500 ms to run https://explain.depesz.com/s/4Acp. I think a chunk of time is spent on this join onto `build_resource_config_version_inputs` in order to remove any builds that have the `resource_id` as an input. Just wanted to mention it in case we can think of anything to refactor this query in order to not need us to do that where clause!

clarafu commented on atc/db/resource.go:
```diff
@@ -953,3 +956,243 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+		SELECT DISTINCT i.build_id
+			FROM build_resource_config_version_inputs i
+			WHERE i.resource_id=$1 AND i.version_md5=$2
+	UNION ALL
```
nit: maybe we can do just a UNION instead of a UNION ALL? Probably because we don't need duplicate resulting build_ids right

clarafu reviewed (commented):
Super duper cool :D

aoldershaw commented on atc/api/resourceserver/versionserver/downstream.go:
```diff
@@ -0,0 +1,56 @@
+		w.Header().Set("Content-Type", "application/json")
+		err = json.NewEncoder(w).Encode(causality)
+		if err != nil {
+			logger.Error("failed-to-encode", err, lager.Data{"resource-name": resourceName, "resource-config-version": versionID})
```
According to the ATC logs, this resource version (https://ci.concourse-ci.org/teams/contributor/pipelines/prs/resources/concourse-pr/causality/939850057/downstream?vars.branch=%22master%22) contains a cycle, causing this encoding step to fail:

```
{"timestamp":"2021-06-08T13:42:14.522065439Z","level":"error","source":"atc","message":"atc.causality.failed-to-encode","data":{"error":"json: unsupported value: encountered a cycle via []*atc.CausalityBuild","resource-config-version":939850057,"resource-name":"concourse-pr","session":"4720233"}}
```

chenbh replied:
Ah, this is caused by the `put` step not generating a new version, it just outputs the same version as the `get` step

chenbh replied:
Ended up flattening the api response so it's no longer a tree. This has the side effect of allowing full cycles in the graph (in addition to the "partial" cycles for resources that bump themselves).

Commit:
address review comments

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
stop passing around list of build ids

..and rerun the recursive query instead. `JOIN` seems to be much faster
than `WHERE IN`, so it should get better performance even if we're
running the recursive query twice.

now that the list of builds is resolved twice, everything needs to be
wrapped in a transaction to prevent race conditions in case a build
completes in between the 2 queries.

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
use both resource and version id to identify rv

since build outputs are uniquely identified by the resource id and
version md5

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

chenbh commented on atc/db/resource.go:
```diff
@@ -953,3 +956,243 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+			return err
+		}
+
+		rv, found := resourceVersions[rcvID]
```
Good catch!

clarafu commented:
Hmm do you think we should remove the `view all` button or tab for the `outputs of` (or for `inputs of` too) if it does not have any outputs/inputs for the user to view? This is because at first glance I thought the `view all` button was for the `metadata` section because it is really close to it and there was no divider between the `metadata` and `outputs of` section when there are no rows within the `outputs of` section. Honestly this is not a big deal though, just something I noticed! 

![Screen Shot 2021-06-09 at 10 12 00 AM](https://user-images.githubusercontent.com/19290703/121371291-70d89f00-c90b-11eb-9489-34131f8a306a.png)

Commit:
inline upstream/downstream logic

in order to support the edge case where a build outputs the same
resource version as one of its inputs, the child of a build node cannot
be the same as the parent. in order to make this work, certain
guarantees around how the tree is constructed needs to be made.

if we can guarantee that the tree is constructed from the leaf up, we
are able to simply remove the resource version from its children's
children. however, this still leaves the problem of how to look at
grandchildren

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

vito committed:
B: flatten causality api response

the old tree-style response runs into problems with the json marshaller
when it tries to encode cyclic resource versions (builds that output the
same version as one of the inputs).

what this commit does is completely flatten the tree, and pre-process
the jobs and resources so that it's easier to build the graph

Signed-off-by: Alex Suraci <suraci.alex@gmail.com>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

vito committed:
B: add endpoint to query versionedResoure by id

we always had this api endpoint, just never needed to use it until now.
it's added because we navigate to the causality page via the version id,
but the breadcrumbs need to display the version string.

also removed the leftover stuff from the previous causality spike

Signed-off-by: Alex Suraci <suraci.alex@gmail.com>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

vito committed:
B: update web to handle flat api response

followup to the previous api change where the causality endpoint now
returns 4 arrays instead of a tree

Signed-off-by: Alex Suraci <suraci.alex@gmail.com>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
fix elm dependencies

these changes were generated by https://github.com/zwilias/elm-json

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented on atc/db/resource.go:
```diff
@@ -949,3 +1000,252 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+			build.ResourceVersionIDs = append(build.ResourceVersionIDs, rv.ID)
+		}
+
+		if used, found := resourceHasParent[rv.ResourceID]; !used || !found {
+			resourceHasParent[rv.ResourceID] = false
+		}
```
Pretty sure this can just be removed - given that if a key isn't found in a `map`, the zero value is returned, the condition should be equivalent to:

```go
if !resourceHasParent[rv.ResourceID] {
	resourceHasParent[rv.ResourceID] = false
}
```

...which is redundant. We can just not store resources in the map if they don't have a parent for the same result.

aoldershaw commented on web/public/causality.mjs:
```diff
@@ -0,0 +1,49 @@
+  graphviz.graphviz.layout(dot, "svg", "dot").then(content => {
+    svg.html(content);
+
+    svg.selectAll('title').remove();
+    svg.selectAll('*').attr('xlink:title', null);
```
Could we add a comment explaining why we need this? IIRC it was to get rid of the HTML tooltips that can't be disabled with graphviz

aoldershaw commented on web/elm/src/Views/TopBar.elm:
```diff
@@ -194,17 +208,56 @@ jobBreadcrumb jobName =
+                    Assets.UpstreamCausalityComponent
+
+        name =
+            String.join "," <| Concourse.versionQuery version
```
Wonder if we should do a similar truncation for resource version components here - otherwise, the breadcrumb can be really long and look a bit wonky
<img width="1904" alt="Screen Shot 2021-06-11 at 4 01 50 PM" src="https://user-images.githubusercontent.com/26583442/121742937-8ba74100-cace-11eb-872a-af5b0849ca60.png">

chenbh replied:
I feel like any non-css based solutions here would have trouble based on the size of the users device. We were able to use it in the SVG because the user can pan around, and we're only concerned with the minimum number of character to uniquely identify a resource.

Tried to solve this in https://github.com/concourse/concourse/pull/7125/commits/6786808e5d03fa53b2f8df8bbc41e0f7daf54a7f, but that only solves the problem if the very last breadcrumb overflows. Which should be fine if we assume pipeline and resource names to be somewhat short, and the only long breadcrumbs are the resource versions

aoldershaw replied:
Maybe we could take the approach that we tried to take initially with the resource versions and wrap each key-value pair in a `span` that has a `max-width` and applies ellipsis to text if it overflows. So, the version breadcrumb may end up looking like:

```
digest:sha256:abc123foobar..., tag:branch-resource-causal...
```

Not a perfect solution as it can still wrap to multiple lines, though.

Commit:
B: load resource inputs/outputs immediately

before it would wait until the 5 second clock timer before it refreshes
the page and fetches the "inputs to" and "outputs of" sections. now we
fetch it as soon as we verify it's a valid resource version

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
code cleanup and better documentation

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

Commit:
B: topbar enables text-overflow on last breadcrumb

by allowing the last breadcrumb component to grow, we can use css to
properly compute overflows. this would still break if some previous
breadcrumb (e.g. pipeline name) is so long that the other components
don't even show up.

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
S: move subpage ordering around

it makes more sense that the causality page comes before the
pseudo-pages like NotFound and FlySuccess

Signed-off-by: Bohan Chen <bochen@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
B: switch to version string base causality

it only seems to do badly on super large datasets, and we're planning to
limit the max number of builds/input/outputs in a followup commit

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw committed:
set max number of builds/rvs in causality graph

currently set to arbitrary values 5000 and 25000 based on our very
limited poking into ci's db.

we used 403 forbidden since it was the closest status that could be
argued for: the application logic determined that nobody should have
access to this resource as it would be too expensive to compute.

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>
Co-authored-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented on atc/db/resource.go:
```diff
@@ -949,3 +993,263 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
+}
+
+func causalityQuery(direction CausalityDirection) string {
+	switch direction {
```
I notice you deployed the updated version to CI. Performance doesn't seem too bad with the version_md5 based approach!

I haven't tried to reason about why this is happening yet (i.e. whether it's a bug with the UI or the API endpoints), but it seems like the graph is rendered in the wrong direction - e.g. when looking at the downstream graph, the resource version appears on the right of the graph with the job edges feeding into it, rather than out of it (e.g. https://ci.concourse-ci.org/teams/main/pipelines/concourse/resources/concourse/causality/999241219/downstream):

<img width="1277" alt="Screen Shot 2021-06-25 at 3 32 06 PM" src="https://user-images.githubusercontent.com/26583442/123476300-82d46600-d5ca-11eb-9faf-9279d05f2e8b.png">

chenbh replied:
This, along with some other subtle bugs were fixed in https://github.com/concourse/concourse/pull/7125/commits/ff4a1a0c750471907aab693928b72630c029d5df

Commit:
B: filter out unused resource version from api res

it can be argued that having the extra resource_version_id doesn't hurt
- the ui handles missing nodes anyways. but it felt more responsible to
  properly sanitize the response anyways

in addition, i'm not sure when but the tree building (i.e.
handleinput/handleoutput) was reversed

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
B: add help text for when causality graph is empty

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
fix causality graph zoom not resetting

i think one of the differences between the causality and pipeline graph
is that the pipeline graph needs to retain the zoom/translate if the
pipeline config gets updated, no such concerns is needed for the
causality graph

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
fixup tests

Signed-off-by: Bohan Chen <bochen@pivotal.io>

chenbh commented:
@aoldershaw deployed the latest build on CI, there is an interesting case with jobs that bump their own resources. The graph for https://ci.concourse-ci.org/teams/main/pipelines/concourse/resources/concourse/causality/1002942545/downstream shows 2 versions for the root resource `concourse` because of the `bosh-bump #533`. Because it `get` and `put` to the same resource, the previous version of `concourse-release-repo` was caused by the previous commit of `concourse`, leading it to show up multiple times.

Not sure how to address this, it might be confusing and takes a bit of time to track it down, but it does display the correct info and that could be more important to figure out edge cases like this one.

aoldershaw commented on web/public/causality.mjs:
```diff
@@ -0,0 +1,48 @@
+        var ev = d3.event;
+        g.attr("transform", "translate(" + ev.translate + ") scale(" + ev.scale + ")");
+      })
+    );
```
Minor annoyance, but if you zoom/pan on one page, then go to another causality page and try to zoom/pan, it jumps back to the previous state:

https://user-images.githubusercontent.com/26583442/123835347-6a1fc500-d8d6-11eb-8cd0-5bb2e8de7b57.mov

This can be fixed pretty easily here:

```diff
diff --git a/web/public/causality.mjs b/web/public/causality.mjs
index 87b1de232..e980dca77 100644
--- a/web/public/causality.mjs
+++ b/web/public/causality.mjs
@@ -43,6 +43,8 @@ function createSvg(svg) {
         g.attr("transform", "translate(" + ev.translate + ") scale(" + ev.scale + ")");
       })
     );
+    // Reset graph focus in case we previously modified it on another page.
+    zoom.translate([0,0]).scale(1).center(0,0);
   }
   return g;
 }
```

chenbh replied:
This should be fixed with https://github.com/concourse/concourse/commit/d453ff233f45aa8d68fb76bcd583859fb5e0c002#diff-2e1332d21e1a3253c1661d71595bca330a08e4722f22a8f9127be4101537bfccR23 and I don't see it when I do it, is this after hard reloading the page?

aoldershaw replied:
Ah yeah i guess i had an old version 👍

aoldershaw reviewed (commented):
Have one tiny suggestion about fixing the jumpiness when you zoom/pan and navigate to another causality graph, but otherwise I think it's pretty much good to go.

I don't think the "2 root versions" thing is such an issue, because as you say, it does display the correct information (albeit in a slightly confusing way). We could just prevent adding more resource versions to the root version, but not sure if that's necessarily the right approach

aoldershaw reviewed (approved):
In that case, I'm good to approve. Merge at your own will!

Commit:
S: combine causality endpoints into single file

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
add feature flag for resource causality

since the causality query is so expensive, it makes sense to give
cluster admins the ability to disable the endpoints and page view.

this is the first time we introduced the idea of feature flags to the
frontend. my solution is to include the current state of the feature
flags in the `/api/v1/info` endpoint, this can then be parsed by the
frontend and stored in `session` making it available to any elm code
that might require it.

if resource causality is disabled (default), the `view all` buttons in
the expended resource version will not show up and trying to navigate to
the causality pages directly via url will result in a standard 404 not
found page.

another change is the stauts code for the causality endpoint will be 403
Forbidden if causality is disabled, and 422 UnprocessableEntity if the
graph is too large.

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
fixup tests

Signed-off-by: Bohan Chen <bochen@pivotal.io>

Commit:
fix yarn analyse

it really didn't like it if a record only has a single field, so let's
just add all the fields even though the UI currently only cares about
`resourceCausality`

Signed-off-by: Bohan Chen <bochen@pivotal.io>

chenbh commented:
I've decided to gate the api endpoint and the web UI behind a feature flag mainly so that a cluster with many curious users doesn't kill the database (although the better solution could've been to rate limit the endpoint). Most of it is contained in https://github.com/concourse/concourse/pull/7125/commits/b9d824ee2150d7609cc4d7a82b9254ca77676ee1

Commit:
fixup more tests

Signed-off-by: Bohan Chen <bochen@pivotal.io>

aoldershaw commented on atc/api/infoserver/info.go:
```diff
@@ -15,6 +15,15 @@ func (s *Server) Info(w http.ResponseWriter, r *http.Request) {
 		WorkerVersion: s.workerVersion,
 		ExternalURL:   s.externalURL,
 		ClusterName:   s.clusterName,
+		FeatureFlags: map[string]bool{
```
Very sensible approach putting it in the `info` response, but I'll propose the idea that first came to my mind. What if instead we interpolate the (relevant) flags into the elm flags directly when we interpolate `index.html`:
https://github.com/concourse/concourse/blob/49b742951de35f289a5d752f6aa17a6d211f0a6a/web/public/index.html#L73-L79

If we did it this way, we wouldn't need to wait for a full round trip to the ATC to make decisions. This isn't such a big deal here (the causality buttons may take a split second to load if you load the resource page directly), but waiting for a round trip isn't ideal for features like #5802.

Anyway, not a big deal, happy to keep it how to have it - the API-based approach could be used in places like `fly` as well, so has its own benefits.

chenbh replied:
Yeah, the main problem with this approach is that the feature flags are not known at subpage init time. I wonder if we can have some sort of pre-init that blocks until `ClusterInfo` has been fetched. But that's an idea for another time

aoldershaw replied:
I think rather than baking more into Elm with a pre-init or something, an approach like I suggested might be better. But anyway, for now, I think the current approach is A-OK

aoldershaw reviewed (approved):
Going to re-approve but looks like there are some merge conflicts now 😬
